### DISCUSSION TITLE
having `magit-file-track`

### DISCUSSION THREAD

**Discussion by mmahmoudian** [magit/magit#5115] [Ideas]
### General idea

I'm proposing to have a function to add untracked files to the git. The name comes from the existing function named `magit-file-untrack`.

I imagine it works by asking the user to pick a file (similar to `C-x C-f`) and then staging the file.

### [My] Use-case

This function comes handy when user have hidden untracked files (the only way to tame a bare repo especially when used for managing dotfiles):

```sh
git config --local status.showUntrackedFiles no
```

When the untracked files are hidden, there is no way in magit to stage an untracked file, other than jumping to a terminal and use `git add path/file` and then get to magit and `magit-refresh`.

**Comment by kyleam** [comment]
Have you tried `magit-stage-file`?

https://github.com/magit/magit/blob/0e8f25a8d8011328f2bf082232c720b24c2a12c2/lisp/magit-apply.el#L328-L347

**Reply by mmahmoudian** [reply]
That will not work because as I explained the files are not tracked, meaning they are not displayed, and pressing <kbd>s</kbd> or `M-x magit-stage-file` does not let me to select a file from outside of magit interface.

Am I missing anything?

**Reply by tarsius** [reply]
> Am I missing anything?

Yes, this part:

> With prefix argument FORCE, offer ignored files for completion."

In other words, try `C-u M-x magit-stage-file`.

**Reply by mmahmoudian** [reply]
Thanks. I tried it before my previous comment, and I tried it again now too, but it does not work or at least I cannot comprehend how to use it. Here is the process:
1. `C-u C-x magit-stage-file`
2. the minibuffer shows `Stage ignored file,s:` and it waits for an input
3. I write a path (e.g `/home/meh`) and I press tab for completion, but it only shows `[No matches]`:
    ![image](https://github.com/magit/magit/assets/390889/8e088083-dbcf-4a7c-8462-dd7d4f7d2133)
4. I clear the whole text and press tab. This time only bring 154 files, all of which are from folders in the root of the git repo that start with `.` (i.e matching some of the folders that match `/^\..*/` regex), although in that folder I have many many more files and folders that are starting with `.`.
    ![image](https://github.com/magit/magit/assets/390889/8d031cd9-037c-495b-88dc-49de40e51b65)


In this repo, I do not have `.gitignore` file, but instead all the untracked files are ignored (as I briefly explained in the original post). 

My magit config is bare-bone:

```lisp
(use-package magit
  :ensure t
  :defer t)
```

Am I missing something?

**Reply by tarsius** [reply]
This command uses functions, which use `git ls-files`, to list files. That git command is not affected by `status.showUntrackedFiles`. Since these files are not *ignored* (using the mechanism described in gitignore(5)), they are merely untracked. So they are listed when *not* using a prefix argument.

In one of the screenshots I can see that you are using an absolute path. That won't work because this command does not use a `find-file`-like mechanism. That would be harder to implement. Instead it compiles a list of completion candidates beforehand and then hands that to `completing-read`. The members of that list are file names *relative to the root of the git working tree*. So if you begin with `/home`, then nothing will be matched.

(In the second screenshot, I miss seeing either the `Stage ignored file,s: ` or `Stage file,s: ` prompt. You seem to have invoked something else.)

**Reply by mmahmoudian** [reply]
@tarsius Thanks for your patience :) I know there is a higher chance that I'm missing something rather than magit.

> In one of the screenshots I can see that you are using an absolute path.

Thanks for the response. But I tried it with absolute and also local path and t didn't work

> (In the second screenshot, I miss seeing either the Stage ignored file,s: or Stage file,s: prompt. You seem to have invoked something else.)

No, I'm invoking `magit-stage-file`. The second screenshot is after pressing tab hoping for completion (first sentence of step 4)

But to solve all these complications, I think the best is to have a test environment to demo the issue. create an empty folder and run this to have a git with 1 commit and 2 committed files, and 2 untracked files:

```sh
#!/usr/bin/env bash

# create some sample files
touch file1 file2 .file3 .file4
# create a git repo here
git init
# hide the untracked files
git config --local status.showUntrackedFiles no
# stage one file of each type
git add file1 .file3
# commit staged files
git commit --message='1st commit'

git log

git ls-files
```

Then try `C-u C-x magit-stage-file` and try staging `file2` or `.file4`.

This is what I see in my setup (keycast is enabled in the tab bar at right-top corner):

https://github.com/magit/magit/assets/390889/08d8dd50-bc12-45a5-9f5a-9b3629329c5a


Thanks for your time in advance :)

**Reply by tarsius** [reply]
> Then try `C-u C-x magit-stage-file` and try staging `file2` or `.file4`.

Since these files are *not* "gitignored", but merely hidden from `git status`[^1], you have to use the "unstaged or untracked" variant of `magit-stage-file`. In other words: *no* prefix argument.

For me `M-x magit-stage-file` lists `.file4`, `file2` and `kickstart.sh`.

If it still doesn't for you, then test what the various `magit-*-files` functions, that are used in `magit-state-file`'s interactive form, return. `magit-untracked-files` is supposed to return the files, I just mentioned.

[^1]: and `magit-status`, which also respects `status.showUntrackedFiles`, but that had to be implemented explicitly.

**Reply by mmahmoudian** [reply]
Thanks. As you mentioned, it brings the files in my example environment although it does not work in bare repos. Also There is a practical issue: the `magit-untracked-files` returns which is used in `magit-stage-file` tries to list all the untracked files in the directory (as expected), although it will consume lots of IO when the directory is large (e.g in my case the my whole `~/`). In this case I believe a process akin to `find-file` is more fitting, but as you stated before it "would be harder to implement". I understand that it might not worth the effort.

**One last question**: Is there any upper-limit to the list returned by `(magit-untracked-files)`? because the `fd --type=f --hidden --search-path=/home/mehrad | wc -l` returns `376378` but the following returns only `359551`:

```lisp
(+ (length (magit-untracked-files))
   (length (magit-ignored-files))
   (length (magit-unstaged-files))
   (length (magit-list-files)))
```

which is 16827 files different. The reason this caught my eye is because I have a file named `.toprc` which is not listed by `magit-stage-file`, and I tried to understand the code and wrote this to check (it is find by :

```lisp
(member ".toprc" (magit-untracked-files))
```
which results in "no output":
> ```
> Code block produced no output (took 1.6s).
> ```

but `fd` finds it in the same git repo root:

```console
❯  fd --type=f --hidden '.toprc'
.config/htop/htoprc
.toprc
```

**Reply by tarsius** [reply]
> In this case I believe a process akin to `find-file` is more fitting, but as you stated before it "would be harder to implement". I understand that it might not worth the effort.

It's something for the backlog.

> Is there any upper-limit to the list returned by `(magit-untracked-files)`? because the `fd --type=f --hidden --search-path=/home/mehrad | wc -l` returns `376378` but the following returns only `359551`:

Magit does not impose any limit, but `git ls-files` might. I don't know.

**Reply by mmahmoudian** [reply]
> It's something for the backlog.

Nice, exciting times ahead then 🤗 I might even get my elisp-foo string enough to make a PR (highly unlikely though as I don't know any lisp and I'm working on learning scheme).

> Magit does not impose any limit, but git ls-files might. I don't know

Well, my `git ls-file | wc -l` is 40, but because my bare repo is on my home folder, there are a ton of untracked files. So my guess was some kind of limitation on `magit-untracked-files`

**Reply by tarsius** [reply]
>> It's something for the backlog.

> Nice, exciting times ahead then

I am tracking this in a private todo list. I am closing this because I just read through all this once more to get to the quoted part, and would like to avoid doing that again in a few months.
